docs/internal/rfc/009rfc-transition-explorer.md

```diff
@@ -0,0 +1,459 @@
+type Model = Either[UninitializedModel, InitializedModel]
+
+/**  The type of errors that can result from failures when loading a spec. */
+type LoadErr
+
+/**  The type of errors that can result from failures to assert a constraint. */
+type AssertErr
+
+/**  The type of errors that can result from checking satisfiability of a frame. */
+type SatError
+
+/**  Maintains session and connection information */
+type Connection
+
+trait TransEx {
+
+  /** Used internally */
+  private def connection(): Connection
+
+  /**  Create a transition explorer
+   *
+   * Establishes the connection and a running session
+   *
+   * The channel is managed by the gRPC framework. */
+  def apply(channel: Channel): TransEx
+
+
+  /** Reset the state of the explorer
+   *
+   * Returns the explorer to the same state as when the currently loaded model
+   * was freshly loaded. Used to restart exploration from a clean slate.
+   *
+   * [TRANS-EX.1::LOAD.1]
+   */
+  def reset(): Unit
+
+  /** Load a model for exploration
+   *
+   * If a model is already loaded, it will be replaced and the state of the exploration
+   * [[reset]].
+   *
+   * [TRANS-EX.1::QCHECK.1]
+   * [TRANS-EX.1::LOAD.1]
+   * [TRANS-EX.1::SBMC.1::TRANSITIONS.1]
+   * 
+   * @param spec the root TLA+ module 
+   * @param aux auxiliary modules that may be required by the root module
+   * @return `Left(LoadErr)` if parsing or loading the model from `spec` goes
+   *          wrong, or `Right(UninitializedModel)` if the model is loaded successfully.
+   */
+  def loadModel(spec: String, aux: List[String] = List()): Either[LoadErr, UninitializedModel]
+
+  /** Initialize the constants of a model
+   *
+   * This will always also reset an exploration back to its initial frame. */
+  def initializeConstants(constants: Map[String, TlaEx]): Either[AssertErr, InitializedModel]
+
+  /** Prepare the loaded modle with the given `transition` */
+  def prepareTransition(transition: Transition): Either[AssertErr, Unit]
+
+  /** The transitions that have been prepared. */
+  def preparedTransitions(): Set[Transition]
+
+  /** Apply a (previously prepared) transition to the current frame.
+   *
+   * Without any arguments, a previously prepared transition is selected
+   * nondeterministically.
+   *
+   * When given a `transition`, apply it if it is already prepared, or prepare
+   * it and then apply it, if not.
+   * 
+   * Interfaces to `assumeTransition` and `pickTransition`, followed by
+   * `nextState`. */
+  def applyTransition(): Either[AssertErr, Unit]
+  def applyTransition(transition: Transition): Either[AssertErr, Unit]
+
+  /** Assert a constraint into the current frame
+   *
+   *  Interface to `assertState` */
+  def assert(assertion: TlaEx): Either[AssertErr, Unit]
+
+  /** The example of an execution from the an initial state up to the current symbolic state
+   *
+   * TODO: I don't know how it is expected to obtain multile different executions. */
```

Comment:
We will be able to enumerate multiple transitions in a loop like:
  - `sat`
  - `execution`
  - `assert(<the execution is different from the last one>)`
  - `sat`
  - `execution`
  - ...

There are many ways to exclude executions. One of them is by using the view mapping.

Reply:
How will the user form the assertion `<the execution is different from the last one>`?

Reply:
For instance, by adding an assertion that `tx.type` is different from the types we have in the last execution.

Reply:
This poses the question, which we havne't talked about yet, of how users will form constraints. Do you imagine that to be in the form of an EDSL to construct our IR, or as just raw strings that we'll parse server side?

Reply:
I expect the users to use Apalache IR in the json format. This part is well documented, so it makes sense to use it.

Reply:
Very good. So then the libraries we provide for users will probably include a module for constructing the IR. I'll clarify this TODO, and then I think we're good to merge?

Reply:
I also think we are good to merge. I am not sure how we should go about IR. It is a very simple tree-like representation, which makes JSON (or s-expressions) quite natural there. Do we want to mirror this simple representation with gRPC objects?

Reply:
Added a note on this. Agree a JSON ser/de (and s-expers for the better languages ;P) will be a good fit. I think we won't need to mirror this gRPC, we'll just delegate to the ser/de for the AST representation in the target language in the libs we provide. I've added a note on this.